### PR TITLE
sql: make `SHOW LAST QUERY STATISTICS` more compatible across versions

### PR DESCRIPTION
**Pull request by knz**
Requested/suggested by @bdarnell in [this comment](https://github.com/cockroachdb/cockroach/pull/67853#issuecomment-884231205).
Follows up to the discussion in #67618.
First commit from #67919.

Prior to this patch, the statement was always using a fixed set of
columns. The set of columns would be modified (usually extended) every
time we wanted to give insight into a new timing.

This was problematic, as we really wish to use the statement using
clients built in a different version of CockroachDB. Previous-version
clients usually do not know about future-version columns, and
next-version clients get confused by not receiving data from columns
they expected.

To improve upon this situation, this patch takes inspiration from Go's
own `metrics` package's API: the client of the API spells out which
particular counters they wish to receive a value for. Only the
specified counters are returned, and if a counter is not recognized
its returned value is "undefined". In any case, no error is reported,
which simplifies the client code.

No release note since this observer statement is not documented.

Release note: None

### CODE REVIEW

**Review by otan — COMMENTED** [review]
any reason not to use `SELECT col1, col2 FROM [SHOW LAST QUERY STATISTICS]`?

it means we can start straightaway with this.

**Comment by knz** [comment]
It's an observer statement, it cannot be part of a sub query.

File: pkg/cli/clisqlclient/conn.go [hunk 1]
```diff
@@ -191,17 +193,39 @@ func (c *sqlConn) EnsureConn() error {
...
-	_, _, _, _, _, _, err := c.getLastQueryStatisticsInternal()
-	if err != nil {
-		fmt.Fprintf(c.errw, "warning: cannot show server execution timings: %v\n", err)
-		c.connCtx.EnableServerExecutionTimings = false
-	} else {
+	// Starting in v21.2 servers, clients can request an explicit set of
+	// values which makes them compatible with any post-21.2 column
+	// additions.
+	_, err := c.QueryRow("SHOW LAST QUERY STATISTICS RETURNING x", nil)
+	if err == nil {
```
Comment (otan) [inline pkg/cli/clisqlclient/conn.go:212]: should we catch for syntax errors / do retries?
a transient error may disable query statistics for the entire session. maybe we want to log that.

**Review by otan — APPROVED** [review]

**Review by knz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67917)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @otan)

___
*[pkg/cli/clisqlclient/conn.go, line 212 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/67917#-MfDPWBVh4hllw-xpFd8-r3-212:-MfX1EbJ-hnwE_BiOF_f:b-c30u1x) ([raw file](https://github.com/cockroachdb/cockroach/blob/72d28608f245bdc382f4e4aa09cab7d663c21616/pkg/cli/clisqlclient/conn.go#L212)):*
retries can't happen with observer statements.
But you're right, let's check that this only matches syntax errors. Done

>  a transient error may disable query statistics for the entire session. maybe we want to log that.

I added a warning message, but I'm not too concerned about the "stickiness" given that I'm still planning to get #66258 merged.

**Comment by knz** [comment]
TFYR!

bors r=otan

**Pull request merged**